governor/src/gcra.rs

```diff
@@ -111,6 +111,42 @@ impl Gcra {
         let tau = self.tau;
         let t = self.t;
         state.measure_and_replace(key, |tat| {
+            let tat = tat.unwrap_or_else(|| self.starting_state(t0));
+            let earliest_time = tat.saturating_sub(tau);
+            if t0 < earliest_time {
+                Err(MW::disallow(
+                    key,
+                    StateSnapshot::new(self.t, self.tau, earliest_time, earliest_time),
+                    start,
+                ))
+            } else {
+                let next = cmp::max(tat, t0) + t;
+                Ok((
+                    MW::allow(
+                        key,
+                        StateSnapshot::new(self.t, self.tau, t0, next)),
+                        next,
+                ))
+            }
+        })
+    }
+
+    pub(crate) fn peek_test<
+        K,
+        P: clock::Reference,
+        S: StateStore<Key = K>,
+        MW: RateLimitingMiddleware<P>,
+    >(
+        &self,
+        start: P,
+        key: &K,
+        state: &S,
+        t0: P,
+    ) -> Result<MW::PositiveOutcome, MW::NegativeOutcome> {
+        let t0 = t0.duration_since(start);
+        let tau = self.tau;
+        let t = self.t;
+        match state.measure_and_peek(key, |tat| {
```

Comment:
Hmmm, I had a feeling that it should be possible to use `measure_and_replace` by having the closure return `MW::disallow` and doing some manual math, in https://github.com/antifuchs/governor/compare/fix-120. This doesn't really work though, so duplicating that code and getting it right, there, seems better (:

Reply:
Sorry, for the delayed response, I ran stable formatting now.